File: backend/app/services/dashboard_service.py
from __future__ import annotations

from collections import Counter
from datetime import datetime, timedelta, timezone

from app.config import AppSettings
from app.models import (
    ActivityItemModel,
    AgentModel,
    DashboardMetricModel,
    DashboardResponse,
    GraphEdgeModel,
    GraphNodeModel,
    InventoryResponse,
    OrganizationChartResponse,
    OverviewModel,
    RouteModel,
    RouterGraphResponse,
    SkillModel,
)
from app.services.config_reader import CodexConfigReader


class DashboardService:
    """
    summary: 설정 스캔 결과와 활동성 집계를 API 응답 모델로 변환한다.
    purpose/context: UI가 조직도와 대시보드를 즉시 렌더링할 수 있도록 서버 측 조합 로직을 담당한다.
    rationale: Codex CLI의 로컬 설정(TOML/JSON)은 파편화되어 있으므로, 이를 하나의 일관된 인벤토리와 
               그래프 모델로 통합하여 프런트엔드의 복잡도를 낮추고 데이터 일관성을 보장한다.
    input: 로컬 Codex 설정을 읽는 reader와 전역 설정을 주입받는다.
    output: overview, inventory, router graph, organization chart, dashboard 응답을 생성한다.
    rules/constraints: 엔진 타입(Codex/Gemini)에 따라 조직도와 에이전트 목록을 동적으로 구성한다.
    failure behavior: 데이터가 비어 있어도 빈 컬렉션과 0값을 반환해 UI가 실패하지 않게 한다.
    """

    def __init__(self, reader: CodexConfigReader, settings: AppSettings) -> None:
        self._reader = reader
        self._settings = settings
        self._founder_name = settings.founder_name.strip() or "대표이사"

    def build_inventory(self, engine: str | None = None) -> InventoryResponse:
        """
        사용 가능한 모든 스킬과 에이전트 정보를 집계하고 상호 연결 상태를 검증한다.
        에이전트가 참조하는 스킬이 실제로 존재하는지, 라우터에 등록되어 있는지 등을 체크하여
        시스템의 '건강 상태'를 진단하는 핵심 로직임.
        """
        target_engine = engine or self._settings.default_engine
        skills_raw = self._reader.read_skills(target_engine)
        agents_raw = self._reader.read_agents(target_engine)
        router_config = self._reader.read_router_config(target_engine)
        enabled_paths = self._reader.read_enabled_skill_paths()
        routes = self._extract_routes(router_config)
        routed_agent_names = {route.agent_name for route in routes}

        skills = [
            SkillModel(
                name=skill["name"],
                path=skill["path"],
                installed=True,
                enabled=skill["path"] in enabled_paths,
            )
            for skill in skills_raw
        ]
        skill_name_set = {skill.name for skill in skills}
        skill_path_set = {skill.path for skill in skills}
        skill_by_name = {skill.name: skill for skill in skills}
        skill_by_path = {skill.path: skill for skill in skills}

        agents = []
        for agent in agents_raw:
            agent_name = str(agent.get("name", "unknown-agent"))
            description = str(agent.get("description", ""))
            short_description = self._to_optional_str(agent.get("short_description")) or self._short_description(description)
            one_click_prompt = self._to_optional_str(agent.get("one_click_prompt"))
            raw_skill_name = self._to_optional_str(agent.get("skill_name"))
            raw_skill_path = self._to_optional_str(agent.get("skill_path"))
            skill_name, skill_path, mapping_note = self._resolve_skill_mapping(
                raw_skill_name,
                raw_skill_path,
                skill_by_name,
                skill_by_path,
            )
            department_label_ko = self._to_optional_str(agent.get("department")) or self._settings.default_department_label_ko
            role_label_ko = self._to_optional_str(agent.get("role_label")) or self._settings.default_role_label_ko
            is_routed = agent_name in routed_agent_names
            status, reason = self._resolve_agent_status(skill_name, skill_path, skill_name_set, skill_path_set, is_routed)
            if mapping_note:
                reason = f"{reason} {mapping_note}"
            agents.append(
                AgentModel(
                    name=agent_name,
                    role_label_ko=role_label_ko,
                    department_label_ko=department_label_ko,
                    description=description,
                    short_description=short_description,
                    one_click_prompt=one_click_prompt,
                    skill_name=skill_name,
                    skill_path=skill_path,
                    routing_type=str(agent.get("routing_type", "unknown")),
                    routed=is_routed,
                    status=status,
                    reason=reason,
                )
            )

        return InventoryResponse(skills=skills, agents=agents, routes=routes)

    def build_overview(self, engine: str | None = None) -> OverviewModel:
        target_engine = engine or self._settings.default_engine
        inventory = self.build_inventory(target_engine)
        return OverviewModel(
            total_skills=len(inventory.skills),
            total_agents=len(inventory.agents),
            routed_agents=len([a for a in inventory.agents if a.routed]),
            route_hints=len(inventory.routes),
            broken_mappings=len([a for a in inventory.agents if a.status == "broken"]),
            active_threads=len(self._reader.read_recent_threads(100, target_engine)),
            active_agents=len([a for a in inventory.agents if a.status == "healthy"]),
            last_scanned_at=self._reader.get_scan_timestamp(),
        )

    def build_router_graph(self, engine: str | None = None) -> RouterGraphResponse:
        """
        라우터와 에이전트 간의 관계를 그래프 형태로 변환한다.
        사용자가 어떤 키워드를 입력했을 때 어떤 에이전트가 반응하는지를 시각적으로 파악할 수 있도록 돕는다.
        """
        target_engine = engine or self._settings.default_engine
        router_config = self._reader.read_router_config(target_engine)
        routes = self._extract_routes(router_config)
        inventory = self.build_inventory(target_engine)
        agent_map = {a.name: a for a in inventory.agents}

        nodes = [GraphNodeModel(id="router", label="Router", type="router", status="healthy")]
        edges = []
        seen_agents = set()

        for route in routes:
            agent_name = route.agent_name
            if agent_name not in seen_agents:
                label = agent_name
                agent_info = agent_map.get(agent_name)
                status = "healthy"
                if agent_info:
                    label = f"{agent_info.role_label_ko}\n({agent_name})"
                    status = agent_info.status
                nodes.append(GraphNodeModel(id=agent_name, label=label, type="agent", status=status))
                seen_agents.add(agent_name)
            edges.append(GraphEdgeModel(id=f"e-{agent_name}", source="router", target=agent_name, label=route.keyword))

        return RouterGraphResponse(nodes=nodes, edges=edges)

    def build_org_chart(self, engine: str | None = None) -> OrganizationChartResponse:
        """
        부서-에이전트 계층 구조를 트리 형태의 그래프 데이터로 구성한다.
        에이전트의 소속 부서 정보를 기준으로 노드를 그룹화하여 조직도를 생성함.
        """
        target_engine = engine or self._settings.default_engine
        inventory = self.build_inventory(target_engine)
        nodes = [GraphNodeModel(id="founder", label=self._founder_name, type="founder", status="healthy")]
        edges = []

        departments = sorted(list({a.department_label_ko for a in inventory.agents}))
        for dept in departments:
            nodes.append(GraphNodeModel(id=dept, label=dept, type="department", status="healthy"))
            edges.append(GraphEdgeModel(id=f"e-{dept}", source="founder", target=dept))

            dept_agents = [a for a in inventory.agents if a.department_label_ko == dept]
            for agent in dept_agents:
                label = f"{agent.role_label_ko}\n({agent.name})"
                nodes.append(GraphNodeModel(id=agent.name, label=label, type="agent", status=agent.status))
                edges.append(GraphEdgeModel(id=f"e-{agent.name}", source=dept, target=agent.name))

        return OrganizationChartResponse(nodes=nodes, edges=edges)

    def build_dashboard(self, engine: str | None = None) -> DashboardResponse:
        """
        시스템 활동 이력을 분석하여 주요 지표와 타임라인 데이터를 생성한다.
        최근 실행 횟수와 트렌드를 계산하여 대시보드 요약 정보를 제공함.
        """
        target_engine = engine or self._settings.default_engine
        history = self._reader.read_history(target_engine)
        now = datetime.now(timezone.utc)
        window_start = now - timedelta(days=self._settings.trend_window_days)

        recent_history = [
            item
            for item in history
            if self._parse_iso(item.get("timestamp")) and self._parse_iso(item.get("timestamp")) > window_start
        ]

        total_exec_count = len(history)
        recent_exec_count = len(recent_history)

        agent_usage = Counter([item.get("agent") for item in history if item.get("agent")])
        
        # Trend calculation
        trend_data = self._calculate_trend(recent_history, window_start, now)

        metrics = [
            DashboardMetricModel(key="total_exec", label="전체 실행 횟수", value=total_exec_count, trend_values=trend_data),
            DashboardMetricModel(key="recent_exec", label="최근 실행", value=recent_exec_count, trend_values=[]),
            DashboardMetricModel(key="active_agents", label="활성 에이전트", value=len(agent_usage), trend_values=[]),
        ]

        activities = [
            ActivityItemModel(
                title=str(item.get("agent", "unknown")),
                subtitle=str(item.get("prompt", ""))[:60],
                timestamp=self._parse_iso(item.get("timestamp")) or now,
            )
            for item in history[-10:]
        ]
        activities.reverse()

        return DashboardResponse(
            metrics=metrics,
            active_agents=activities, # 단순 매핑
            recent_skills=[],
            recent_threads=[],
            timeline=activities,
            department_breakdown=[],
            status_breakdown=[],
        )

    def _extract_routes(self, router_config: dict) -> list[RouteModel]:
        routes = []
        raw_routes = router_config.get("routes", [])
        if isinstance(raw_routes, list):
            for r in raw_routes:
                routes.append(
                    RouteModel(
                        agent_name=str(r.get("agent", "unknown")),
                        keyword=str(r.get("intent", ""))[:30],
                    )
                )
        return routes

    def _resolve_skill_mapping(
        self,
        raw_name: str | None,
        raw_path: str | None,
        skill_by_name: dict[str, SkillModel],
        skill_by_path: dict[str, SkillModel],
    ) -> tuple[str | None, str | None, str | None]:
        if raw_path:
            skill = skill_by_path.get(raw_path)
            if skill:
                return skill.name, skill.path, None
            return None, raw_path, "(경로 불일치)"
        if raw_name:
            skill = skill_by_name.get(raw_name)
            if skill:
                return skill.name, skill.path, None
            return raw_name, None, "(이름 불일치)"
        return None, None, None

    def _resolve_agent_status(
        self,
        skill_name: str | None,
        skill_path: str | None,
        skill_name_set: set[str],
        skill_path_set: set[str],
        is_routed: bool,
    ) -> tuple[HealthStatus, str]:
        if not skill_name and not skill_path:
            return "broken", "연결된 스킬 정보가 없습니다."
        if skill_path and skill_path not in skill_path_set:
            return "broken", f"스킬 경로를 찾을 수 없습니다: {skill_path}"
        if skill_name and skill_name not in skill_name_set:
            return "broken", f"스킬 이름이 유효하지 않습니다: {skill_name}"
        if not is_routed:
            return "passive", "라우터에 등록되지 않은 에이전트입니다."
        return "healthy", "정상 작동 중입니다."

    def _calculate_trend(self, history: list[dict], start: datetime, end: datetime) -> list[int]:
        buckets = self._settings.trend_buckets
        delta = (end - start) / buckets
        trend = []
        for i in range(buckets):
            b_start = start + delta * i
            b_end = b_start + delta
            count = len([
                item for item in history
                if b_start <= (self._parse_iso(item.get("timestamp")) or start) < b_end
            ])
            trend.append(count)
        return trend

    @staticmethod
    def _parse_iso(value: object) -> datetime | None:
        if not isinstance(value, str):
            if isinstance(value, datetime):
                return value
            return None
        try:
            return datetime.fromisoformat(value.replace("Z", "+00:00"))
        except (ValueError, AttributeError):
            return None

    @staticmethod
    def _to_optional_str(value: object) -> str | None:
        if value is None:
            return None
        s = str(value).strip()
        return s if s else None

    @staticmethod
    def _short_description(text: str) -> str:
        s = text.strip().split("\n")[0]
        if len(s) > 80:
            return f"{s[:77]}..."
        return s

File: electron/services/DashboardService.ts
/**
 * DashboardService — 대시보드 데이터 집계 서비스.
 *
 * @what
 * - 실행 통계(stats), 최근 활동(recentActivity), 인벤토리(inventory), 개요(overview),
 *   라우터 그래프(routerGraph), 조직도(orgChart) 등 대시보드 화면에 필요한 데이터를
 *   ConfigReader로부터 읽어 가공합니다.
 *
 * @design
 * - 모든 응답은 Zod 스키마(ipc-contract)로 검증되어 타입 안전성을 보장합니다.
 * - 에이전트 상태(healthy/broken/passive)와 스킬 매핑 상태를 실시간으로 분석합니다.
 * - 데이터 변환 책임을 ConfigReader(읽기)와 분리하여 단일 책임 원칙을 따릅니다.
 *
 * @usage
 *   const svc = new DashboardService(configReader);
 *   const stats = svc.getStats(runList);
 *   const overview = svc.getOverview('gemini');
 */
import { z } from 'zod';
import {
  InventoryResponseSchema, OverviewResponseSchema,
  RouterGraphResponseSchema, OrganizationChartResponseSchema,
} from '../../types/ipc-contract';
import { ConfigReader } from './ConfigReader';
import { SETTINGS } from '../settings/AppSettings';

/** 대시보드 최근 활동 목록의 단일 항목을 나타냅니다. */
export interface DashboardActivity {
  /** 활동 ID */
  id: string;
  /** 에이전트 ID */
  agentId: string;
  /** 활동 상태 (completed, failed, running 등) */
  status: string;
  /** 활동 설명 메시지 */
  message: string;
  /** 활동 발생 타임스탬프 */
  timestamp: string;
}

/**
 * 라우터 설정에서 키워드-에이전트 매핑 라우트 목록을 추출합니다.
 * routes 배열과 routing_hints 객체를 모두 처리합니다.
 * @param routerConfig - 라우터 설정 객체
 * @returns 키워드와 에이전트 이름 쌍 배열
 */
function _extractRoutes(routerConfig: Record<string, unknown>): Array<{ keyword: string; agentName: string }> {
  const routes: Array<{ keyword: string; agentName: string }> = [];
  const rawRoutes = routerConfig.routes;
  if (Array.isArray(rawRoutes)) {
    for (const r of rawRoutes) {
      if (r && typeof r === 'object') {
        routes.push({ agentName: String((r as Record<string, unknown>).agent || 'unknown'), keyword: String((r as Record<string, unknown>).intent || '').slice(0, 30) });
      }
    }
  }
  const hints = routerConfig.routing_hints;
  if (hints && typeof hints === 'object') {
    for (const [keyword, agentName] of Object.entries(hints as Record<string, unknown>)) {
      routes.push({ keyword: keyword.slice(0, 30), agentName: String(agentName) });
    }
  }
  return routes;
}

/**
 * 에이전트 레코드의 스킬 매핑 정보를 기반으로 상태를 판별합니다.
 * @param agent - 에이전트 설정 객체
 * @returns 'healthy' 또는 'broken' 상태 문자열
 */
function _resolveStatusSimple(agent: Record<string, unknown>): string {
  const skillName = agent.skill_name ? String(agent.skill_name) : null;
  const skillPath = agent.skill_path ? String(agent.skill_path) : null;
  const broken = agent.broken === true;
  if (broken || (!skillName && !skillPath)) return 'broken';
  return 'healthy';
}

export class DashboardService {
  /** 설정 데이터 읽기를 위임받은 ConfigReader 인스턴스 */
  private configReader: ConfigReader;

  /**
   * DashboardService 인스턴스를 생성합니다.
   * @param configReader - 데이터 읽기에 사용할 ConfigReader
   */
  constructor(configReader: ConfigReader) {
    this.configReader = configReader;
  }

  /**
   * 실행 목록에서 통계를 집계합니다.
   * @param runList - 실행 레코드 배열
   * @returns 총 실행 수, 에이전트 수, 업타임, 성공률, 완료/실패 수
   */
  getStats(runList: Array<{ status: string }>): { totalRuns: number; totalAgents: number; uptime: number; successRate: number; completedRuns: number; failedRuns: number } {
    const configStats = this.configReader.getStats();
    const completedRuns = runList.filter(r => r.status === 'completed').length;
    const totalRuns = runList.length;
    const successRate = totalRuns > 0 ? Math.round((completedRuns / totalRuns) * 100) : 0;
    return { totalRuns, totalAgents: configStats.totalAgents, uptime: configStats.uptime, successRate, completedRuns, failedRuns: runList.filter(r => r.status === 'failed').length };
  }

  /**
   * 최근 실행 목록에서 대시보드 표시용 활동 항목을 생성합니다.
   * @param runs - 실행 레코드 배열
   * @returns 최대 10개의 DashboardActivity 항목
   */
  getRecentActivity(runs: Array<Record<string, unknown>>): DashboardActivity[] {
    return runs.slice(0, 10).map(run => ({
      id: run.id as string, agentId: run.agentId as string, status: run.status as string,
      message: run.status === 'completed' ? `${run.id} completed` : run.status === 'failed' ? `${run.id} failed${run.error ? `: ${run.error}` : ''}` : `${run.id} ${run.status}`,
      timestamp: run.createdAt as string,
    }));
  }

  /**
   * 지정된 엔진의 스킬, 에이전트, 라우트 정보를 통합한 인벤토리 데이터를 반환합니다.
   * 에이전트-스킬 매핑 상태(healthy/broken/passive)도 함께 분석합니다.
   * @param engine - 엔진 이름 (선택)
   * @returns Zod 스키마로 검증된 인벤토리 응답
   */
  getInventory(engine?: string): Record<string, unknown> {
    const targetEngine = typeof engine === 'string' ? engine : undefined;
    const skills = this.configReader.readSkills(targetEngine).map(s => ({ name: s.name, path: s.path, installed: true, enabled: true }));
    const agentsRaw = this.configReader.readAgents(targetEngine);
    const routerConfig = this.configReader.readRouterConfig(targetEngine);
    const enabledPaths = this.configReader.readEnabledSkillPaths();
    const routes = _extractRoutes(routerConfig);
    const routedNames = new Set(routes.map(r => r.agentName));
    const skillNames = new Map(skills.map(s => [s.name, s]));
    const skillPaths = new Map(skills.map(s => [s.path, s]));

    const agents = agentsRaw.map((a: Record<string, unknown>) => {
      const name = String(a.name || 'unknown');
      const desc = String(a.description || '');
      const rawSkillName = a.skill_name ? String(a.skill_name) : null;
      const rawSkillPath = a.skill_path ? String(a.skill_path) : null;

      let resolvedName: string | null = rawSkillName;
      let resolvedPath: string | null = rawSkillPath;
      let mappingNote: string | null = null;
      if (rawSkillPath && skillPaths.has(rawSkillPath)) {
        resolvedName = skillPaths.get(rawSkillPath)!.name;
        resolvedPath = rawSkillPath;
      } else if (rawSkillPath) {
        resolvedPath = null;
        mappingNote = '(경로 불일치)';
      } else if (rawSkillName && skillNames.has(rawSkillName)) {
        resolvedName = rawSkillName;
        resolvedPath = skillNames.get(rawSkillName)!.path;
      } else if (rawSkillName) {
        resolvedName = null;
        mappingNote = '(이름 불일치)';
      }

      const isRouted = routedNames.has(name);
      let status: string; let reason: string;
      if (!resolvedName && !resolvedPath) { status = 'broken'; reason = '연결된 스킬 정보가 없습니다.'; }
      else if (resolvedPath && !skillPaths.has(resolvedPath)) { status = 'broken'; reason = `스킬 경로를 찾을 수 없습니다: ${resolvedPath}`; }
      else if (resolvedName && !skillNames.has(resolvedName)) { status = 'broken'; reason = `스킬 이름이 유효하지 않습니다: ${resolvedName}`; }
      else if (!isRouted) { status = 'passive'; reason = '라우터에 등록되지 않은 에이전트입니다.'; }
      else { status = 'healthy'; reason = '정상 작동 중입니다.'; }
      if (mappingNote) reason = `${reason} ${mappingNote}`;

      return {
        name,
        roleLabelKo: String(a.role_label || SETTINGS.founderName),
        departmentLabelKo: String(a.department || '관리지원'),
        description: desc,
        shortDescription: a.short_description ? String(a.short_description) : (desc.split('\n')[0]?.slice(0, 80) ?? ''),
        oneClickPrompt: a.one_click_prompt ? String(a.one_click_prompt) : null,
        skillName: resolvedName,
        skillPath: resolvedPath,
        routingType: String(a.routing_type || 'unknown'),
        routed: isRouted,
        status, reason,
      };
    });

    const result = { skills, agents, routes };
    return safeValidate(InventoryResponseSchema, result);
  }

  /**
   * 지정된 엔진의 개요(overview) 데이터를 집계합니다.
   * 스킬/에이전트 수, 라우팅 현황, 활성 스레드, 손상된 매핑 등을 포함합니다.
   * @param engine - 엔진 이름 (선택)
   * @returns Zod 스키마로 검증된 개요 응답
   */
  getOverview(engine?: string): Record<string, unknown> {
    const targetEngine = typeof engine === 'string' ? engine : undefined;
    const skills = this.configReader.readSkills(targetEngine);
    const agentsRaw = this.configReader.readAgents(targetEngine);
    const routerConfig = this.configReader.readRouterConfig(targetEngine);
    const routes = _extractRoutes(routerConfig);
    const activeThreads = this.configReader.readRecentThreads(100, targetEngine).length;
    const brokenCount = agentsRaw.filter((a: Record<string, unknown>) => {
      const rawSkillName = a.skill_name ? String(a.skill_name) : null;
      const skillPath = a.skill_path ? String(a.skill_path) : null;
      return !rawSkillName && !skillPath;
    }).length;
    const routedCount = routes.length;
    const healthyCount = agentsRaw.length - brokenCount;

    const result = {
      totalSkills: skills.length, totalAgents: agentsRaw.length,
      routedAgents: routedCount, routeHints: routes.length,
      brokenMappings: brokenCount, activeThreads,
      activeAgents: healthyCount, lastScannedAt: new Date().toISOString(),
    };
    return safeValidate(OverviewResponseSchema, result);
  }

  /**
   * 라우터-에이전트 간 연결 관계를 그래프(노드/엣지) 형태로 반환합니다.
   * @param engine - 엔진 이름 (선택)
   * @returns 노드와 엣지 배열 (Zod 검증됨)
   */
  getRouterGraph(engine?: string): { nodes: Array<Record<string, unknown>>; edges: Array<Record<string, unknown>> } {
    const targetEngine = typeof engine === 'string' ? engine : undefined;
    const routerConfig = this.configReader.readRouterConfig(targetEngine);
    const routes = _extractRoutes(routerConfig);
    const agentsRaw = this.configReader.readAgents(targetEngine);
    const agentMap = new Map(agentsRaw.map((a: Record<string, unknown>) => [String(a.name), a]));

    const nodes: Array<Record<string, unknown>> = [{ id: 'router', label: 'Router', type: 'router', status: 'healthy' }];
    const edges: Array<Record<string, unknown>> = [];
    const seenAgents = new Set<string>();

    for (const route of routes) {
      if (!seenAgents.has(route.agentName)) {
        const agent = agentMap.get(route.agentName);
        const label = agent ? `${agent.role_label || route.agentName}\n(${route.agentName})` : route.agentName;
        const status = agent ? _resolveStatusSimple(agent) : 'passive';
        nodes.push({ id: route.agentName, label, type: 'agent', status });
        seenAgents.add(route.agentName);
      }
      edges.push({ id: `e-${route.agentName}`, source: 'router', target: route.agentName, label: route.keyword });
    }

    const result = { nodes, edges };
    return safeValidate(RouterGraphResponseSchema, result);
  }

  /**
   * 창업자-부서-에이전트 계층 구조의 조직도를 그래프 형태로 반환합니다.
   * @param engine - 엔진 이름 (선택)
   * @returns 노드와 엣지 배열 (Zod 검증됨)
   */
  getOrgChart(engine?: string): { nodes: Array<Record<string, unknown>>; edges: Array<Record<string, unknown>> } {
    const targetEngine = typeof engine === 'string' ? engine : undefined;
    const agentsRaw = this.configReader.readAgents(targetEngine);

    const nodes: Array<Record<string, unknown>> = [{ id: 'founder', label: SETTINGS.founderName, type: 'founder', status: 'healthy' }];
    const edges: Array<Record<string, unknown>> = [];
    const deptSet = new Set<string>();

    for (const a of agentsRaw) {
      const dept = String(a.department || '관리지원');
      if (!deptSet.has(dept)) {
        nodes.push({ id: dept, label: dept, type: 'department', status: 'healthy' });
        edges.push({ id: `e-${dept}`, source: 'founder', target: dept });
        deptSet.add(dept);
      }
      const name = String(a.name || 'unknown');
      const roleLabel = String(a.role_label || SETTINGS.founderName);
      const status = _resolveStatusSimple(a);
      nodes.push({ id: name, label: `${roleLabel}\n(${name})`, type: 'agent', status });
      edges.push({ id: `e-${name}`, source: dept, target: name });
    }

    const result = { nodes, edges };
    return safeValidate(OrganizationChartResponseSchema, result);
  }
}

/**
 * Zod 스키마로 데이터를 검증하고 실패 시 상세한 검증 오류를 throw합니다.
 * @param schema - 검증에 사용할 Zod 스키마
 * @param data - 검증할 데이터
 * @returns 검증된 타입 안전 데이터
 */
function safeValidate<T>(schema: z.ZodType<T>, data: unknown): T {
  try {
    return schema.parse(data);
  } catch (err) {
    if (err instanceof z.ZodError) {
      const issues = err.issues.map(i => `${i.path.join('.')}: ${i.message}`).join(', ');
      throw new Error(`Validation failed: ${issues}`);
    }
    throw err;
  }
}
